datalad_next/url_operations/ssh.py

```diff
@@ -160,50 +162,59 @@ def download(self,
-            # download can start
-            self._progress_report_start(
-                progress_id,
-                ('Download %s to %s', from_url, to_path),
-                'downloading',
-                expected_size,
-            )
-            for chunk in stream:
-                # write data
-                dst_fp_write(chunk)
-                # compute hash simultaneously
-                hasher.update(chunk)
-                self._progress_report_update(
-                    progress_id, ('Downloaded chunk',), len(chunk))
-            props.update(hasher.get_hexdigest())
-            return props
-        except CommandError as e:
-            if e.code == 244:
-                # this is the special code for a file-not-found
-                raise UrlOperationsResourceUnknown(from_url) from e
-            else:
-                # wrap this into the datalad-standard, but keep the
-                # original exception linked
-                raise UrlOperationsRemoteError(from_url, message=str(e)) from e
+            with iter_subproc(cmd) as stream:
+                props = self._get_props(from_url, stream)
+                expected_size = props['content-length']
+                # The stream might have changed due to not yet processed, but
+                # fetched data, that is now chained in front of it. Therefore we
+                # get the updated stream from the props
+                download_stream = props.pop('_stream')
+
+                dst_fp = sys.stdout.buffer \
+                    if to_path is None \
+                    else open(to_path, 'wb')
+
+                # Localize variable access to minimize overhead
+                dst_fp_write = dst_fp.write
+
+                # download can start
+                for chunk in side_effect(
+                        lambda chunk: self._progress_report_update(
+                            progress_id,
+                            ('Downloaded chunk',),
+                            len(chunk)
+                        ),
+                        download_stream,
+                        before=partial(
+                            self._progress_report_start,
+                            progress_id,
+                            ('Download %s to %s', from_url, to_path),
+                            'downloading',
+                            expected_size
+                        )
+                ):
+                    # write data
+                    dst_fp_write(chunk)
+                    # compute hash simultaneously
+                    hasher.update(chunk)
+
+        except IterableSubprocessError as e:
+            self._check_return_code(e.returncode, from_url)
+        except StopIteration:
+            raise UrlOperationsResourceUnknown(from_url)
         finally:
-            if dst_fp and to_path is not None:
-                dst_fp.close()
             self._progress_report_stop(progress_id, ('Finished download',))
```

Comment:
How confident are you that it is OK to skip closing `dst_fp`?

It would eventually be closed on garbage collection, I guess. But the timing of that is relatively unpredictable. What made you remove it?